File: frontend/src/types/habit.ts
export interface Habit {
  id: number;
  name: string;
  description?: string | null;
  frequency: 'daily' | 'weekly' | 'monthly';
  category: 'health' | 'work' | 'learning' | 'other';
  start_date: string;
  user_id: number;
}

export interface HabitCheckin {
  id: number;
  habit_id: number;
  checkin_date: string;
  notes?: string | null;
  status: string;
}

export interface HabitWithCheckins extends Habit {
  checkins: HabitCheckin[];
  streak: number;
  completedToday: boolean;
}

export type HabitCategory = 'health' | 'work' | 'learning' | 'other';
export type HabitFrequency = 'daily' | 'weekly';

export const HABIT_CATEGORIES: { value: HabitCategory; label: string; emoji: string }[] = [
  { value: 'health', label: 'Health', emoji: '🏃' },
  { value: 'work', label: 'Work', emoji: '💼' },
  { value: 'learning', label: 'Learning', emoji: '📚' },
  { value: 'other', label: 'Other', emoji: '✨' },
];

export const HABIT_FREQUENCIES: { value: HabitFrequency; label: string }[] = [
  { value: 'daily', label: 'Daily' },
  { value: 'weekly', label: 'Weekly' },
];

File: backend/models/habit.py
from sqlalchemy import Column, String, Integer, ForeignKey, Date, Enum
from sqlalchemy.orm import relationship
from backend.models.base import BaseModel

class Habit(BaseModel):
    __tablename__ = "habits"
    name = Column(String, index=True)
    description = Column(String, nullable=True)
    frequency = Column(Enum("daily", "weekly", "monthly", name="frequency_types"))
    target = Column(Integer, nullable=True)
    user_id = Column(Integer, ForeignKey("users.id"))

    user = relationship("User", back_populates="habits")
    checkins = relationship("HabitCheckin", back_populates="habit", cascade="all, delete-orphan")

class HabitCheckin(BaseModel):
    __tablename__ = "habit_checkins"
    checkin_date = Column(Date, index=True)
    status = Column(String, default="completed")
    notes = Column(String, nullable=True)
    habit_id = Column(Integer, ForeignKey("habits.id"))

    habit = relationship("Habit", back_populates="checkins")